service/storage_service.cc

```diff
@@ -1390,10 +1504,28 @@ std::unordered_set<locator::token> storage_service::get_tokens_for(inet_address
 }
 
 // Runs inside seastar::async context
-void storage_service::set_gossip_tokens(const std::unordered_set<dht::token>& local_tokens) {
+// Assumes that no other functions modify STREAMS_TIMESTAMP, TOKENS or STATUS
+// in the gossiper's local application state while this function runs.
+void storage_service::gossip_joined(
+        const std::unordered_set<dht::token>& tokens, std::optional<db_clock::time_point> streams_ts) {
+    assert(!tokens.empty());
+
+    if (streams_ts) {
+        // Order is important: both the streams timestamp and tokens must be known when a node handles our status.
+        _gossiper.add_local_application_state({
+            { gms::application_state::TOKENS, value_factory.tokens(tokens) },
+            { gms::application_state::STREAMS_TIMESTAMP, value_factory.streams_timestamp(*streams_ts) },
+            { gms::application_state::STATUS, value_factory.normal(tokens) }
+        }).get();
+        return;
+    }
+
+    // !streams_ts => we are upgrading from a non-CDC version.
+
+    // Order is important: tokens must be known when a node handles our status.
     _gossiper.add_local_application_state({
-        { gms::application_state::TOKENS, value_factory.tokens(local_tokens) },
-        { gms::application_state::STATUS, value_factory.normal(local_tokens) }
+        { gms::application_state::TOKENS, value_factory.tokens(tokens) },
```

Comment:
> According to your "order is important comment", we would need to set STREAMS_TIMESTAMP first (i.e., te new function) and only then TOKENS and STATUS (the pre-existing function), but that's easy, isn't it?

First, that would introduce more complex relationships and invariants in the code ("function A must be called before function B or it won't work"), and I think I introduced enough of these already (not counting the existing relationships...)

Second, I'm not sure if it's enough. This variant of `add_local_application_state` which receives multiple values was introduced to fix a bug:
https://github.com/scylladb/scylla/issues/2894
I'm not sure if I understand the problem description, but apparently, in the following code:
```
    gossiper.add_local_application_state(gms::application_state::TOKENS, value_factory.tokens(local_tokens)).get();
    gossiper.add_local_application_state(gms::application_state::STATUS, value_factory.normal(local_tokens)).get();
```
the TOKENS value might receive a _higher_ version than the STATUS value, reach some node without STATUS, and then, when STATUS eventually reaches the node, it will be ignored.
I don't see how TOKENS can get a higher version than STATUS though. @tgrabiec maybe you could explain?

Anyway, now we have the "atomic":
```
 _gossiper.add_local_application_state({
  { gms::application_state::TOKENS, value_factory.tokens(tokens) },
  { gms::application_state::STATUS, value_factory.normal(tokens) }
}).get();
```

Reply:
I'm afraid I simply don't understand this issue to make any intelligent comment about it, so I'll resolve this conversation. It seems you and @tgrabiec already gave this issue enough thought.